backend/src/pong/pong.gateway.ts

```
@@ -10,117 +11,142 @@ import {
+	}
+
+	@SubscribeMessage('joinGameRoom')
 	joinGameRoom(@MessageBody() game: Game, @ConnectedSocket() client: Socket) {
+		console.log('CurretnGame: ', game);
 		const cookie: string = jwtCookieFromHandshakeString(
 			client.handshake.headers.cookie,
 		);
 		try {
 			const userId: number = this.authService.userIdFromCookieString(cookie);
-			if (userId === game.player_one.id) {
-				this.playerOneIsInGame = true;
-				this.playerOneId = game.player_one.id;
-			}
-			else if (userId === game.player_two.id) {
-				this.playerTwoIsInGame = true;
-			}
-			const roomName = this.buildUniqueRoomId(game);
-			client.join(roomName);
-			this.gameState.roomName = roomName;
-			if (this.playerOneIsInGame && this.playerTwoIsInGame) {
-				this.server.in(roomName).emit("GameCanStart");
+			if (game.id && game.player_one && game.player_two) {
+				if (!gameSubscribers[game.id]) {
+					this.createGameInstance(game, client);
+				} else {
+					client.join(gameSubscribers[game.id].roomName);
+				}
+				if (userId === game.player_one.id) {
+					gameSubscribers[game.id].playerOneIsInGame = true;
+					gameSubscribers[game.id].playerOneId = userId;
+				} else if (userId === game.player_two.id) {
+					gameSubscribers[game.id].playerTwoIsInGame = true;
+					gameSubscribers[game.id].playerTwoId = userId;
+				}
+				if (
+					!this.pongService.gameHasStarted &&
+					gameSubscribers[game.id].playerOneIsInGame &&
+					gameSubscribers[game.id].playerTwoIsInGame
+				) {
+					console.log('Emitting game start');
+					this.server
+						.in(gameSubscribers[game.id].roomName)
+						.emit('GameCanStart');
+					this.pongService.gameHasStarted = true;
+				}
 			}
 		} catch (e) {}
 	}
 
-	sendPositionOfElements(@MessageBody() gameState: GameState) {
-		this.server.in(this.gameState.roomName).emit('elementPositions', gameState);
+	sendPositionOfElements(gameState: GameState) {
+		this.server
+			.in(gameSubscribers[gameState.gameId].roomName)
```

should this have a check if gameSubscribers[gameState.gameId] exists?
not sure if it's possible to get here after the room has been destroyed...
also not sure if `this.server.in("non_existant_room").emit()` throws or just does nothing, because if it just silently fails it's not a problem at all

---

oh no.. never mind
than we still would have to protect accessing the `roomName` property of something undefined...

---

In the "controller" on line 124, i check if the room exists, and only then proceed to call the methods that look "unprotected"